acceptance: enhance the Docker tests.

knz opened the pull request:
Needed to investigate  #15510.

Two patches:

###  	acceptance: enhance the collection of docker logs upon Wait()

**tl;dr** output of command inside container now copied to `console-output.log`

###  	cli/interactive_tests: place test output details in the artifacts directory

**tl;dr** tests now place output directly in artifacts.

knz committed:
acceptance: enhance the collection of docker logs upon Wait()

**tl;dr** output of command inside container now copied to `console-output.log`

The stdout/stderr output produced by the command running inside the
container is collected after the container has signalled termination
of the command, in `(*Container).Wait()`.

Prior to this patch, this output would simply be copied to the test's
stderr, which:

- is silenced entirely by Go's `testing` module if the test is
  successful or if `-test.v` was not specified;
- when it is printed out for the user to see, it is interleaved with
  the logging output on the host.

The latter point in particular caused difficulty in investigating
issues: even under *normal* circumstances logging output by Docker is
already very noisy, and when Docker misbehaves the logging output can
then fill with garbage. This makes it difficult to extract any logging
output by the command running inside the container, where the focus in
troubleshooting a failing test should be.

This patch alleviates the issue by guaranteeing that the output of the
command inside the container is also copied to a file
`console-output.log` in that container's logging directory, so that it
can be inspected separately.

knz committed:
cli/interactive_tests: place test output details in the artifacts directory

**tl;dr** tests now place output directly in artifacts.

Prior to this patch, the TCL tests in cli/interactive_tests were
emitting output only on stdout/stderr, where Docker could find this
and pump the data to the host test's stderr.

A problem with this approach is that if Docker fails to complete
execution of the container (e.g. `(*Container).Wait()` fails to see a
valid API result from Docker), that output is simply lost, creating
much unneeded difficulty with troubleshooting the test.

This patch alleviates the issue by ensuring that the TCL tests now
write useful test details to the bound artifact directory `logs`
inside the container (including setting `--log-dir` for all spawned
servers to there) -- ensuring that TC and the user can find this data
even if Docker misbehaves.

(The reason why this was not done earlier is that I simply didn't know
that the artifacts directory was already mounted/bound in `logs`
inside the container.)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15516)
<!-- Reviewable:end -->

knz committed:
envutil, cli: allow the user to override the home directory with env var HOME

Prior to this patch, the CLI shell would determine the user's home
directory directly using package `user`, which bypasses the
environment variables and queries the system's user database directly
(e.g. `/etc/passwd` on Unix). This prevented redirecting the SQL
shell's history to a different directory reliably.

This patch addresses this limitation by introducing a new helper
`envutil.HomeDir()` which first uses `$HOME` if that is set, and
otherwise falls back to `user.Current()` (the previous behavior). The
SQL shell is modified to use that, as well as the (obscure) one other
point in the source code (in the terraform framework) that also needed
to know about the home directory.

knz committed:
acceptance: remove test artifacts when the test is successful

tbg reviewed (approved):
my nits are optional.

tschottdorf commented on pkg/acceptance/util.go:
```diff
@@ -569,7 +569,18 @@ func testDocker(
+		return err
+	}
+	// Clean up the log files if the run was successful.
+	if l.LogDir != "" {
```
Hmm, this would be pretty bad if we shared `l.LogDir` between the tests, which I don't think we do, but running a test where `l.LogDir = $HOME` would be pretty bad. Maybe this is too risky given that it doesn't buy us much?

tschottdorf commented on pkg/acceptance/terrafarm/farmer.go:
```diff
@@ -160,7 +160,7 @@ func (f *Farmer) CollectLogs() {
 		return
 	}
-	if err := os.MkdirAll(f.AbsLogDir(), 0777); err != nil {
+	if err := os.MkdirAll(f.AbsLogDir(), 0777|os.ModeSetuid|os.ModeSetgid); err != nil {
```
nit: might as well shell out a `const dirMode = 0777|os.ModeSetuid|os.ModeSetgid`.

tschottdorf commented on pkg/acceptance/cluster/localcluster.go:
```diff
@@ -517,10 +522,10 @@ func (l *LocalCluster) startNode(ctx context.Context, node *testNode) {
 		dockerlogDir := "/logs/" + node.nodeStr
-		locallogDir = filepath.Join(l.logDir, node.nodeStr)
-		maybePanic(os.MkdirAll(locallogDir, 0777))
+		locallogDir = filepath.Join(l.LogDir, node.nodeStr)
```
nit: might as well `s/locallogDir/localLogDir/` at this point.

knz committed:
acceptance: name oneshot artifact directories after the (sub)test

knz committed:
cli/interactive_tests: deflake test_reconnect

knz commented:
Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/acceptance/util.go, line 578 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15516#-KiuhW-q5RGTeTl7d-wS-r4-578:-Kiz-7hwqph0U5AqpSrQ:b-x7c24h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eb75aa050bbe3c8bcf782e1e88c6c6786013315/pkg/acceptance/util.go#L578)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Hmm, this would be pretty bad if we shared `l.LogDir` between the tests, which I don't think we do, but running a test where `l.LogDir = $HOME` would be pretty bad. Maybe this is too risky given that it doesn't buy us much?
</blockquote></details>

I added a flag to ensure that only directories that were created by the test framework can be removed.

---

*[pkg/acceptance/cluster/localcluster.go, line 527 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15516#-KiuhW-q5RGTeTl7d-wQ-r4-527:-Kiz-EjJ-t6tbkqi90K9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eb75aa050bbe3c8bcf782e1e88c6c6786013315/pkg/acceptance/cluster/localcluster.go#L527)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

nit: might as well `s/locallogDir/localLogDir/` at this point.
</blockquote></details>

Done.

---

*[pkg/acceptance/terrafarm/farmer.go, line 163 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15516#-KiuhW-q5RGTeTl7d-wR-r4-163:-Kiz0Fym_rhLDfu0WMqY:b-j3rawl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5eb75aa050bbe3c8bcf782e1e88c6c6786013315/pkg/acceptance/terrafarm/farmer.go#L163)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

nit: might as well shell out a `const dirMode = 0777|os.ModeSetuid|os.ModeSetgid`.
</blockquote></details>

Actually here the change was unneeded/incorrect because only the local/host process is writing to the directory.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15516)*
<!-- Sent from Reviewable.io -->

knz committed:
cli/interactive_tests: ensure that test_history can be ran out of Docker
